Create selectors for the Live Preview feature

okmttdhr opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/79219, https://github.com/Automattic/wp-calypso/issues/79218, https://github.com/Automattic/wp-calypso/pull/79775

## Proposed Changes

Extract functions added in https://github.com/Automattic/wp-calypso/pull/79775 into selectors. This allows us to reuse functions between #79219 and #79218. 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Simple sites

- Patch D117384-code into your sandbox.
- Go to the Theme Showcase page.
- Click any theme and go to the Theme Detail page.
- See if clicking the "Live Preview" button redirects you to the Site Editor (Block Theme Previews).
	- You will not see the button when a theme is;
		- your active theme
		- is not FullSiteEditing compatible
		- a theme with a static page as a homepage 
		- a 3rd party theme
		- a premium theme
		- a woocommerce theme

Atomic sites 

- Go to the Theme Showcase page.
- Click any theme and go to the Theme Detail page. 
- You must install a theme before using the Live Preview on Atomic sites. pbxlJb-3Uv-p2#known-issues
	- When you're using Calypso, activate it once and then reactivate the original theme. 
	- When you're using the classic view, just install it.
- ~~See if clicking the "Live Preview" button redirects you to the Site Editor (Block Theme Previews).~~ -> I am not yet sure how to patch it to make BTP work (c.f. https://github.com/Automattic/dotcom-forge/issues/3322). We can just check if the "Live Preview" button exists or not. 
	- You will not see the button when a theme is;
		- your active theme
		- is not FullSiteEditing compatible
		- a theme with a static page as a homepage 
		- not installed (not listed in "My Themes")

Please see pbxlJb-3Uv-p2#known-issues about unsupported themes. 
I don't think we should go through everything in this PR since I just extracted some functions from #79219. 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

okmttdhr committed:
Add isLivePreviewSupported

okmttdhr committed:
Use isLivePreviewSupported function

okmttdhr committed:
Extract getLivePreviewUrl

okmttdhr committed:
Avoid using flags in multiple places

okmttdhr committed:
Accept siteId for isFullSiteEditingTheme

arthur791004 commented on client/state/themes/selectors/is-live-preview-supported.js:
```diff
@@ -0,0 +1,139 @@
+ *
+ * @see pbxlJb-3Uv-p2
+ */
+export const isLivePreviewSupported = ( state, themeId, siteId ) => {
```
Not a blocker, but how about just calling it `getIsLivePeviewSupported` as we tend to rename the function when we use it 😅

arthur791004 commented on client/my-sites/theme/live-preview-button/index.jsx:
```diff
@@ -1,140 +1,13 @@
-	// A user doesn't want to preview the active theme.
-	if ( isActive ) {
+export const LivePreviewButton = ( { isLivePreviewSupported, translate, livePreviewUrl } ) => {
+	if ( ! isLivePreviewSupported ) {
```
I think we can select both `isLivePreviewSupported` and `livePreviewUrl` inside this component. What do you think?

taipeicoder replied:
+1

arthur791004 commented on client/my-sites/theme/main.jsx:
```diff
@@ -657,23 +646,11 @@ class ThemeSheet extends Component {
+							isLivePreviewSupported={ isLivePreviewSupported }
+							livePreviewUrl={ livePreviewUrl }
+							translate={ translate }
+						></LivePreviewButton>
```
The button doesn't have children, so it can simply be
```diff
- ></LivePreviewButton>
+ />
```

arthur791004 reviewed (approved):
Nice refactor! Leave minor feedback but they're not blockers 🙂

taipeicoder commented on client/my-sites/theme/live-preview-button/index.jsx:
Not a blocker, but I'd like to encourage you to use Typescript for new files and components introduced in this PR, if possible!

okmttdhr replied:
Will create a follow up PR 😄

taipeicoder commented on client/my-sites/theme/main.jsx:
```diff
@@ -92,12 +92,11 @@ import {
 import { getIsLoadingCart } from 'calypso/state/themes/selectors/get-is-loading-cart';
 import { getBackPath } from 'calypso/state/themes/themes-ui/selectors';
 import { getSelectedSiteId } from 'calypso/state/ui/selectors';
-import { getTheme } from '../../state/themes/selectors';
```
I think we also need to remove
```
import QueryTheme from 'calypso/components/data/query-theme';
```

which was introduced in the last PR, and 

```
<QueryTheme themeId={ this.props.themeId } siteId={ siteId } />
```

okmttdhr replied:
Fixed this in https://github.com/Automattic/wp-calypso/pull/80017/commits/29192892abdfed497185d455326db1d460ff596e, and improved in https://github.com/Automattic/wp-calypso/pull/80017/commits/ebea88e61d90d1b19fea949ad9fc11cb71dbbc12.

taipeicoder commented on client/state/themes/selectors/is-live-preview-supported.js:
```diff
@@ -0,0 +1,139 @@
+ * Check if Live Preview is supported.
+ *
+ * The scenarios where the Live Preview does NOT support;
+ * - On both Simple and Atomic sites;
```
We also need to add a check whether the user is logged in or not. Since the BTP doesn't work while logged out.

okmttdhr committed:
Move selectors inside the child

arthur791004 commented on client/my-sites/theme/live-preview-button/index.jsx:
```diff
@@ -1,140 +1,25 @@
+		isLivePreviewSupported,
+		livePreviewUrl,
+	};
+} )( localize( _LivePreviewButton ) );
```
I guess we can simply use hooks instead of the HOC to get the props here 🤔

okmttdhr committed:
Remove unnecessary <QueryTheme/>

okmttdhr committed:
Use useSelector

okmttdhr committed:
Avoid LivePreviewButton appearing a moment later

okmttdhr committed:
Hide the Live Preview when a user is not logged in

okmttdhr committed:
Sort props alphabetically

taipeicoder commented:
@okmttdhr could you confirm that the `livePreviewUrl` is correct? This is what I'm seeing on my end:

![Screenshot 2023-08-02 at 2 08 45 PM](https://github.com/Automattic/wp-calypso/assets/797888/f83cc850-a477-4784-a21a-cd6d305a16c5)

okmttdhr committed:
Rename a file according to a function name

okmttdhr committed:
Fix the wrong Live Preview URL

Always use siteId to get the URL.
And use sourceSiteId when it prevents Layout Shift.

okmttdhr commented:
>could you confirm that the livePreviewUrl is correct? This is what I'm seeing on my end:

Thanks for pointing that out. I addressed it in https://github.com/Automattic/wp-calypso/pull/80017/commits/c2041139cb25aece6d2cdf4758066b68252e5446.

okmttdhr commented:
I'm merging this PR!

taipeicoder commented on client/state/themes/selectors/get-is-live-preview-supported.js:
```diff
@@ -0,0 +1,145 @@
+
+	// Block Theme Previews need the theme installed on Atomic sites.
+	const isAtomic = isSiteAutomatedTransfer( state, siteId );
+	const isThemeInstalledOnAtomicSite = isAtomic && !! getTheme( state, sourceSiteId, themeId );
```
I think we need to use `siteId` instead of `sourceSiteId` here. I'm getting the following error message on Atomic sites:

![Screenshot 2023-08-02 at 3 27 25 PM](https://github.com/Automattic/wp-calypso/assets/797888/7fe0b0e2-a406-4afb-9642-bbef479cb602)

taipeicoder replied:
Let's open a follow-up PR 🙂

okmttdhr replied:
I'll take a look. Where did you see this error? Maybe on the Block Theme Previews page (= Site Editor)?

taipeicoder replied:
Yep!

okmttdhr replied:
@taipeicoder One more question, on which theme did you see the error, as an example? Thanks! 🙏

taipeicoder replied:
I've opened the issue https://github.com/Automattic/wp-calypso/issues/80133 to track this 🙂
